Content.Server/Magic/MagicSystem.cs

```diff
@@ -0,0 +1,312 @@
+        }
+
+        args.Handled = true;
+    }
+
+    /// <summary>
+    /// Spawns 3 walls in front of the caster in a 3x1/1x3 pattern
+    /// Disappears after a set amount of time
+    /// Allows caster to walk through them freely.
+    /// </summary>
+    /// <param name="args"></param>
+    private void OnForceWallSpell(ForceWallSpellEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        var transform = Transform(args.Performer);
+        var directionPos = transform.Coordinates.Offset(transform.LocalRotation.ToWorldVec().Normalized);
+
+        if (!_mapManager.TryGetGrid(transform.GridID, out var mapGrid))
+            return;
+        if (!directionPos.TryGetTileRef(out var tileReference, EntityManager, _mapManager))
+            return;
+
+        var tileIndex = tileReference.Value.GridIndices;
+        var coords = mapGrid.GridTileToLocal(tileIndex);
+        EntityCoordinates coordsPlus;
+        EntityCoordinates coordsMinus;
+
+        var dir = transform.LocalRotation.GetCardinalDir();
+        switch (dir)
+        {
+            case Direction.North:
+            case Direction.South:
+            {
+                coordsPlus = mapGrid.GridTileToLocal(tileIndex + (1, 0));
+                coordsMinus = mapGrid.GridTileToLocal(tileIndex + (-1, 0));
+                break;
+            }
+            case Direction.East:
+            case Direction.West:
+            {
+                coordsPlus = mapGrid.GridTileToLocal(tileIndex + (0, 1));
+                coordsMinus = mapGrid.GridTileToLocal(tileIndex + (0, -1));
+                break;
+            }
+            default:
+                return;
+        }
+
+        SoundSystem.Play(Filter.Pvs(coords), args.ForceWallSound.GetSound(), AudioParams.Default.WithVolume(args.ForceWallVolume));
+        ForceWallSpawnHelper(args, coords);
+        ForceWallSpawnHelper(args, coordsPlus);
+        ForceWallSpawnHelper(args, coordsMinus);
+
+        args.Handled = true;
+    }
+
+    private void ForceWallSpawnHelper(ForceWallSpellEvent args, EntityCoordinates coordinates)
+    {
+        var forceWall = Spawn(args.WallPrototype, coordinates);
+        var comp = EnsureComp<PreventCollideComponent>(forceWall);
+        comp.Uid = args.Performer;
+    }
+
+    /// <summary>
+    /// Opens all doors within range
+    /// </summary>
+    /// <param name="args"></param>
+    private void OnKnockSpell(KnockSpellEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        //Get the position of the player
+        var transform = Transform(args.Performer);
+        var coords = transform.Coordinates;
+
+        SoundSystem.Play(Filter.Pvs(coords), args.KnockSound.GetSound(), AudioParams.Default.WithVolume(args.KnockVolume));
+
+        //Look for doors and don't open them if they're already open.
+        foreach (var entity in _lookup.GetEntitiesInRange(coords, args.Range))
+        {
```

Comment:
As above, it'd be nice to genericise the logic so if I want to make a spell to open 1 door, or doors in a line, it's a simple change. Then the targeted doors get passed into the opening logic.